src/com/stuypulse/stuylib/math/SLMath.java

```diff
@@ -145,6 +145,128 @@ public static double circular(double x) {
+        }
+
+        // Precondition: x-values should be sorted
+        for(int i = 1; i < x.length; ++i) {
+            if(x[i] < x[i - 1]) {
+                throw new IllegalArgumentException("x-values should be in sorted order");
+            }
+        }
+
+        return formula.calculate(x, y);
+    }
+
+    public static final Integral DEFAULT_INTEGRAL = new Integral.Simpson();
+    /**
+     * Calls integrate(double[] x, double[] y, Integral formula) with default formula Simpson.
+     * 
+     * @param x the x-values of the data points, where (x[i], y[i]) is a point
+     * @param y the y-values of the data points, where (x[i], y[i]) is a point
+     * @return approximation of th eintegral from the given points using the Simpson formula.
+     */
+    public static double integrate(double[] x, double[] y) {
+        return integrate(x, y, DEFAULT_INTEGRAL);
+    }
+
+    /**
+     * Integral is an interface used by integrate() to get the method used for numerical integration.  
+     * It also contains several classes containing various formulas for integration.
+     * 
+     * Riemann: Uses the Riemann sum to approximate the integral. (https://en.wikipedia.org/wiki/Riemann_sum)<br>
+     * Trapezoidal: Uses the trapezoidal rule to approximate the integral. (https://en.wikipedia.org/wiki/Trapezoidal_rule)<br>
+     * Simpson: Uses Composite Simpson's Rule for irregularly spaced data to approximate the integral. (https://en.wikipedia.org/wiki/Simpson%27s_rule)<br>
+     */
+    public interface Integral {
+        /**
+         * Given data points, gives an approximation for the integral.
+         * Arrays x and y must be the same length, because they contain data points.
+         * The x-values must be in sorted order.
+         * 
+         * @param x the x-values of the data points, where (x[i], y[i]) is a point
+         * @param y the y-values of the data points, where (x[i], y[i]) is a point
+         * @return approximation of the integral from given points
+         */
+        public double calculate(double[] x, double[] y);
+
+        public class Riemann implements Integral {
+            public double calculate(double[] x, double[] y) {
+                double riemannsum = 0.0;
+                for(int i = 1; i < x.length; ++i) {
+                    riemannsum += y[i] * (x[i] - x[i - 1]);
+                }
+                return riemannsum;
+            }
+        }
+
+        public class Trapezoidal implements Integral {
+            public double calculate(double[] x, double[] y) {
+                double trapezoidsum = 0.0;
+                for(int i = 1; i < x.length; ++i) {
+                    trapezoidsum += (y[i] + y[i - 1]) * (x[i] - x[i - 1]) / 2;
+                }
+                return trapezoidsum;
+            }
+        }
+
+        public class Simpson implements Integral {
+            public double calculate(double[] x, double[] y) {
```

Comment:
I really lke your implementations, they are clean, pls put your author tag on these,

Reply:
myles helped a bit with recommending how i should structure